components/layout_2020/flow/root.rs

```diff
@@ -223,6 +226,40 @@ impl FragmentTreeRoot {
         stacking_context.build_display_list(builder);
     }
 
+    pub fn get_background_style_for_root_element(&self) -> Option<ServoArc<ComputedValues>> {
+        // https://drafts.csswg.org/css-backgrounds/#special-backgrounds
+        // If the first child element is the root element and doesn't have a background
+        // specified, we should apply the style of the <body> element to it.
+        let root_fragment = match self.children.get(0) {
+            Some(fragment_ref) => fragment_ref.borrow(),
+            None => return None,
+        };
+
+        let root_box_fragment = match *root_fragment {
+            Fragment::Box(ref fragment) if fragment.flags.contains(NodeFlags::IS_HTML_ELEMENT) => {
+                fragment
+            },
+            _ => return None,
+        };
+
+        if root_box_fragment.style.has_nontransparent_background() {
+            return Some(root_box_fragment.style.clone());
+        }
+
+        let first_child_ref = match root_box_fragment.children.get(0) {
```

Comment:
This and the other `.children.get(0)` call above are based on the shape of the box tree. But https://drafts.csswg.org/css-backgrounds/#body-background specifies “root element” and “child element” which are found in the DOM tree. These two trees’ shapes can be different because of `display: none`, `display: contents`, pseudo-elements, etc.

Reply:
Isn't checking the node flags enough here?
  * In the case of display:none there will be no fragment marked with the HTML `NodeFlags` so this method should return `None` as expected.
  * in the case of `display:contents`, this method will not find a Fragment marked with the `BODY` `NodeFlags` so it should return the style of the HTML element. This actually isn't happening properly now (it's returning `None`), but I will fix that an upload a new version of the change.

Reply:
Oh sorry, I missed this part of the spec:

>  if no boxes are generated for the element whose background would be used for the canvas (for example, if the root element has `display: none`), then the canvas background is transparent. 

This also apply if the element whose background would be used has `display: contents`. However that can’t happen on the root element because https://drafts.csswg.org/css-display-3/#transformations turns it into `display: block`. This `cfg` needs to be changed to `#[cfg(any(feature = "servo-layout-2020", feature = "gecko"))]`:

https://github.com/servo/servo/blob/3df65c02fe73a4cce9e391fa1de0b9a4d613c51b/components/style/values/specified/box.rs#L353-L359

This leaves the following "interesting" case:

No background:

* `html { display: none } body { background: green }`
* `body { display: none; background: green }`
* `body { display: contents; background: green }`

Green canvas background (including under `<body>`’s 8px margins from the UA stylesheet):

* `html { display: contents /* computes to block */ } body { background: green }`
* `html::before { content: "" } body { background: green }` (even though the box generated by `<body>` is the *second* child box of that generated by `<html>`)

I believe that to implement this spec, the checks for "is the root element" and "is the first element child of the root element" need to be based on the DOM tree through some `ThreadSafeLayoutNode` method, not based on the box tree.

But then maybe the spec doesn’t match reality and other browsers have interop on something else…

Reply:
Okay. I'll modify the code to be entirely dependent on the DOM tree.